### ISSUE TITLE
Failed to guess column names

### ISSUE THREAD

**Issue by yyamano** [embulk/embulk#284]
Embulk fails to guess column names with the following csv file and configuration.

```
% cat y/csv/dame.csv
magic,description
0XDEADBEEF,word-fill pattern
0xCAFEBABE,java byte code
0xDEAD10CC,dead lock
0xDEADFEED,crash report

% cat y/dame.yml
in:
  type: file
  path_prefix: "y/csv/dame"
out:
  type: stdout
% ./embulk-0.7.3.jar guess y/dame.yml
2015-08-24 02:36:02.130 +0000: Embulk v0.7.3
2015-08-24 02:36:03.662 +0000 [INFO] (guess): Listing local files at directory 'y/csv' filtering filename by prefix 'dame'
2015-08-24 02:36:03.678 +0000 [INFO] (guess): Loading files [y/csv/dame.csv]
2015-08-24 02:36:03.783 +0000 [INFO] (guess): Loaded plugin embulk/guess/gzip from a load path
2015-08-24 02:36:03.795 +0000 [INFO] (guess): Loaded plugin embulk/guess/csv from a load path
in:
  type: file
  path_prefix: y/csv/dame
  parser:
    charset: UTF-8
    newline: CRLF
    type: csv
    delimiter: ','
    quote: '"'                                                                                                                                              
    escape: '"'
    trim_if_not_quoted: false
    skip_header_lines: 0
    allow_extra_columns: false
    allow_optional_columns: false
    columns:
    - {name: c0, type: string}
    - {name: c1, type: string}
out: {type: stdout}
Use -o PATH option to write the guessed config file to a file.
```

The problem comes from https://github.com/embulk/embulk/blob/v0.7.3/lib/embulk/guess/csv.rb#L126

```
header_line = (first_types != other_types && first_types.all? {|t| ["string", "boolean"].include?(t) })
```

if we respect skip_header_lines option at guess time and assume the last line of the header lines as a column line, my problem could be fixed. But the assumption is not always correct. For example, there are multiple header lines, but no column line.

Any thoughts?

**Comment by frsyuki** [comment]
yes, that is a limitation of csv-guess plugin right now.
Does your idea require users to set `skip_header_lines` option to the configuration before guessing? I think it requiring that additional work becomes another limitation.

This is another idea:

* If all column types are string or boolean, check length of values.
* For each `column_index`, compute `var1 = variance(rows.map {|row| row[column_index].length })` and `var2 = variance(rows[1..-1].map {|row| row[column_index].length })`.
* If `var1` is small and `var2` large, assume that the first line is the header line.

**Comment by yyamano** [comment]
> Does your idea require users to set skip_header_lines option to the configuration before guessing? I think requiring additional work becomes another limitation.

Yes. Provide a hint for guess is acceptable to me if it reduces total work.

> This is another idea:

It's an interesting idea. You can find a test program and result at https://gist.github.com/yyamano/505b43147ecabc132507

Could you explain your assumption behind the logic for me? I don't understand it:-(

**Comment by frsyuki** [comment]
We can use variance to know distribution of length of values. If values are fixed-length, variance is 0.0. If variance is small enough, it means length of values is almost same. So, if variance is small, but first line has very different length, we can say the first line is likely column name.

Implemented at #295.

**Comment by yyamano** [comment]
Thank you for the commit, but it seems we have room to improve. I just tried the latest embulk from the repo with http://www.city.fukuoka.lg.jp/data/open/cnt/3/39196/1/P01.csv. I removed the first Japanese header line from the csv file.

```
%  pkg/embulk-0.7.4.jar guess /tmp/t/hoo.yml
2015-09-08 17:10:45.505 +0000: Embulk v0.7.4
2015-09-08 17:10:47.034 +0000 [INFO] (guess): Listing local files at directory '/tmp/t/csv' filtering filename by prefix 'P01'
2015-09-08 17:10:47.041 +0000 [INFO] (guess): Loading files [/tmp/t/csv/P01.csv]
2015-09-08 17:10:47.213 +0000 [INFO] (guess): Loaded plugin embulk/guess/gzip from a load path
2015-09-08 17:10:47.239 +0000 [INFO] (guess): Loaded plugin embulk/guess/csv from a load path
in:
  type: file
  path_prefix: /tmp/t/csv/P01
  parser:
    charset: MS932
    newline: CRLF
    type: csv
    delimiter: ','
    quote: ''''
    trim_if_not_quoted: false
    skip_header_lines: 1
    allow_extra_columns: false
    allow_optional_columns: false
    columns:
    - {name: '83', type: string}
    - {name: '9718', type: long}
    - {name: 早良警察署, type: string}
    - {name: '6', type: long}
    - {name: 814-0006, type: string}
    - {name: 早良区百道1-5-15, type: string}
    - {name: '33.582933', type: double}
    - {name: '130.349081', type: double}
    - {name: 092-847-0110, type: string}
    - {name: 092-847-0110, type: string}
    - {name: '', type: string}
    - {name: 'http://www.police.pref.fukuoka.jp/index.php?type=article&mode=articleList&categoryid=22',
      type: string}
    - {name: '', type: string}
    - {name: '', type: string}
    - {name: '', type: string}
    - {name: '', type: string}
    - {name: '', type: string}
    - {name: '', type: string}
    - {name: '', type: string}
    - {name: '', type: string}
    - {name: '83', type: long}
    - {name: '', type: string}
    - {name: '', type: string}
    - {name: '', type: long}
    - {name: '', type: double}
    - {name: '', type: double}
out: {type: stdout}
```

With 0.7.3:

```
%  pkg/embulk-0.7.3.jar guess /tmp/t/hoo.yml
2015-09-08 17:13:21.242 +0000: Embulk v0.7.3
2015-09-08 17:13:22.827 +0000 [INFO] (guess): Listing local files at directory '/tmp/t/csv' filtering filename by prefix 'P01'
2015-09-08 17:13:22.833 +0000 [INFO] (guess): Loading files [/tmp/t/csv/P01.csv]
2015-09-08 17:13:23.029 +0000 [INFO] (guess): Loaded plugin embulk/guess/gzip from a load path
2015-09-08 17:13:23.044 +0000 [INFO] (guess): Loaded plugin embulk/guess/csv from a load path
in:
  type: file
  path_prefix: /tmp/t/csv/P01
  parser:
    charset: MS932
    newline: CRLF
    type: csv
    delimiter: ','
    quote: ''''
    trim_if_not_quoted: false
    skip_header_lines: 0
    allow_extra_columns: false
    allow_optional_columns: false
    columns:
    - {name: c0, type: string}
    - {name: c1, type: long}
    - {name: c2, type: string}
    - {name: c3, type: long}
    - {name: c4, type: string}
    - {name: c5, type: string}
    - {name: c6, type: double}
    - {name: c7, type: double}
    - {name: c8, type: string}
    - {name: c9, type: string}
    - {name: c10, type: string}
    - {name: c11, type: string}
    - {name: c12, type: string}
    - {name: c13, type: string}
    - {name: c14, type: string}
    - {name: c15, type: string}
    - {name: c16, type: string}
    - {name: c17, type: string}
    - {name: c18, type: string}
    - {name: c19, type: string}
    - {name: c20, type: long}
    - {name: c21, type: string}
    - {name: c22, type: string}
    - {name: c23, type: long}
    - {name: c24, type: double}
    - {name: c25, type: double}
out: {type: stdout}
```

Will try more files tomorrow.
